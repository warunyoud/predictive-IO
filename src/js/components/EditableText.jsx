import _ from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

class EditableText extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      focus: false,
      text: props.text
    }

    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
  }

  componentWillReceiveProps(props) {
    this.setState({ text: props.text });
  }

  onPenClick() {
    document.addEventListener('click', this.handleOutsideClick);
    document.addEventListener('keydown', this.handleKeyPress);
    this.setState({ focus: true });
  }

  unFocusInput() {
    document.removeEventListener('click', this.handleOutsideClick);
    document.removeEventListener('keydown', this.handleKeyPress);
    this.setState({ focus: false });
  }



  handleInputChange(event) {
    const target = event.target;
    this.setState({ text: target.value });
  }

  handleKeyPress(event) {
    if (event.keyCode == 27 || event.keyCode == 13) {
      this.unFocusInput();
      if (event.keyCode == 13 && this.props.handleChangeComplete) {
        this.props.handleChangeComplete(this.state.text);
      } else {
        this.setState({ text: this.props.text });
      }
    }
  }

  handleOutsideClick(event) {
    if (!ReactDOM.findDOMNode(this).contains(event.target)) {
      this.unFocusInput();
      this.setState({ text: this.props.text });
    }
  }

  render() {
    if (this.state.focus) {
      return (
        <input
          autoFocus
          value={ this.state.text }
          onChange={ this.handleInputChange.bind(this) }
          className='editable-text-input'
          type='text'/>
      );
    } else {
      return (
        <h1 className='editable-text'>
          {`${this.props.text} `}
          <span
            onClick={ () => this.onPenClick() }
            className='editable-text-pencil'>
            <i className="material-icons" style={{ color: 'gray' }}> edit </i>
          </span>
        </h1>
      );
    }
  }
}

EditableText.propTypes = {
  text: PropTypes.string.isRequired,
  handleChangeComplete: PropTypes.func,
}

export default EditableText;
