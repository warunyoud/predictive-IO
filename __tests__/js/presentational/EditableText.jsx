import React from 'react';
import EditableText from '../../../src/js/presentational/EditableText';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer
    .create(<EditableText
      text= 'hello'
      handleChangeComplete= { jest.fn() }
      />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
