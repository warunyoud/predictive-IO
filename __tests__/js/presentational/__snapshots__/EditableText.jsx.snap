// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<h1
  className="editable-text"
>
  hello 
  <span
    className="editable-text-pencil"
    onClick={[Function]}
  >
    <i
      className="material-icons"
      style={
        Object {
          "color": "gray",
        }
      }
    >
       edit 
    </i>
  </span>
</h1>
`;
